dumper/lib-ext-hive-metastore/src/main/java/com/google/edwmigration/dumper/ext/hive/metastore/utils/PartitionNameGenerator.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2022-2023 Google LLC
+ * Copyright 2013-2021 CompilerWorks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.edwmigration.dumper.ext.hive.metastore.utils;
+
+import com.google.common.collect.Streams;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility method to generate Hive partition names. */
+public final class PartitionNameGenerator {
+
+  private PartitionNameGenerator() {}
+
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionNameGenerator.class);
+
+  /**
+   * Constructs partition name from the list of partition keys and values. It:
+   *
+   * <ul>
+   *   <li>Escapes special symbols in key and value
+   *   <li>Joins key and value with "="
+   *   <li>Joins key-value pairs with "/"
+   * </ul>
+   *
+   * Resulting partition name has a form similar to "partKey1=partValue1/partKey2=partValue2"
+   *
+   * <p>It copies most of the logic of the original Hive metastore function for creating partition
+   * name: {@link <a
+   * href="https://github.com/apache/hive/blob/rel/release-2.3.6/metastore/src/java/org/apache/hadoop/hive/metastore/Warehouse.java#L315">source</a>}
+   */
+  public static String makePartitionName(List<String> partitionKeys, List<String> partitionValues) {
+    return Streams.zip(
+            partitionKeys.stream(),
+            partitionValues.stream(),
+            PartitionNameGenerator::constructPartitionName)
+        .flatMap(Function.identity())
+        .collect(Collectors.joining("/"));
+  }
+
+  private static Stream<String> constructPartitionName(String partitionKey, String partitionValue) {
```

Comment:
Can't this return a `String` and we drop the `flatMap` above?

Reply:
`Stream<String>` and `Optional<String>` could be used interchangeably in this context, but `Optional::stream` is since Java 9. I don't see it in this repo, but pretty much sure it was mentioned that Java 8 is required:

```diff
-        .flatMap(Function.identity())
+        .flatMap(Optional::stream)
```

As for returning normal String – I will need to filter empty ones. I would prefer going with Optionals if Java 8 is not a requirement anymore though:

```diff
-        .flatMap(Function.identity())
+        .filter(part -> !part.isEmpty())
```

WDYT?

Reply:
Let's stay on Java 8 for the time being.
Since the dumper is probably using lots of null checking we could also have `return null` and `Objects::nonNull` as filter predicate.